amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/RetrievalConfig.java

```diff
@@ -117,17 +127,20 @@ public RetrievalConfig initialPositionInStreamExtended(InitialPositionInStreamEx
     }
 
     public RetrievalFactory retrievalFactory() {
-
         if (retrievalFactory == null) {
             if (retrievalSpecificConfig == null) {
                 retrievalSpecificConfig = new FanOutConfig(kinesisClient())
                         .applicationName(applicationName());
                 retrievalSpecificConfig = appStreamTracker.map(multiStreamTracker -> retrievalSpecificConfig,
-                        streamConfig -> ((FanOutConfig)retrievalSpecificConfig).streamName(streamConfig.streamIdentifier().streamName()));
+                        streamConfig -> ((FanOutConfig) retrievalSpecificConfig).streamName(streamConfig.streamIdentifier().streamName()));
             }
             retrievalFactory = retrievalSpecificConfig.retrievalFactory();
         }
+
+        if (retrievalSpecificConfig != null && customDataFetcherProvider != null) {
```

Comment:
At this point the retrieval specific config will definitely not be null. Since customDataFetcherProvider is only applicable for PollingConfig, what you need here is

```
    public RetrievalFactory retrievalFactory() {

        if (retrievalFactory == null) {
            if (retrievalSpecificConfig == null) {
                retrievalSpecificConfig = new FanOutConfig(kinesisClient())
                        .applicationName(applicationName());
                retrievalSpecificConfig = appStreamTracker.map(multiStreamTracker -> retrievalSpecificConfig,
                        streamConfig -> ((FanOutConfig)retrievalSpecificConfig).streamName(streamConfig.streamIdentifier().streamName()));
            }
            if(customDataFetcherProvider != null) {
                retrievalFactory = retrievalSpecificConfig.retrievalFactory(customDataFetcherProvider);
            } else {
                retrievalFactory = retrievalSpecificConfig.retrievalFactory();
            }
        }
        return retrievalFactory;
    }
```